File: src/Algorithm/LeetCode/LeetCode_503_Next_Greater_Element_II/Solution.java
package Algorithm.LeetCode.LeetCode_503_Next_Greater_Element_II;

import java.util.Arrays;
import java.util.Stack;

class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        Stack<Integer> stack = new Stack<>();
        int[] result = new int[n];

        Arrays.fill(result, -1);

        for (int i = 0; i < 2 * n; i++) {
            while (!stack.isEmpty() && nums[stack.peek()] < nums[i % n]) {
                result[stack.pop()] = nums[i % n];
            }

            stack.push(i % n);
        }

        return result;
    }
}

File: src/Algorithm/LeetCode/LeetCode_503_Next_Greater_Element_II/Solution.py
from typing import List


# https://www.programmercarl.com/0503.下一个更大元素II.html
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [-1] * n
        stack = [0]

        for i in range(1, n * 2):
            while len(stack) != 0 and nums[stack[-1]] < nums[i % n]:
                result[stack[-1]] = nums[i % n]
                stack.pop()
            stack.append(i)

        return result
